### PR TITLE
Improved logging of ODSP error responses

### LINKED ISSUE #6943: facetCodes are not making it to telemetry
Here is what I see in telemetry for ODSP errors:
"facetCodes":"REDACTED (arbitrary object)"

We need to have facetCodes in telemetry. Especially if we are going to remove response (see #6901)

### PR DESCRIPTION
**Pull request by markfields**
Related: #6943 (More comprehensive version of https://github.com/microsoft/FluidFramework/pull/6901)

* Add types modeling the shape of response from ODSP, based on looking at telemetry
* ~Extract the error message from the response and log it to `odspResponseErrorMessage`~
* ~Extract the full list of facets (innermost to outermost joined with " << ") and log it to `odspResponseErrorFacets`~
* ~Rename telemetry field of innermost face to `odspResponseInnermostFacet`~

UPDATE: I went back to the existing telemetry prop names to reduce churn and keep the more (potentially) standard `response` name.  Note that the full list of facet codes are not logged by themselves but they are present in `response`.  So leaving #6943 open for the moment, but I do think it can be closed,  just will leave that determination up to others since I'm about to be AFK.

### CODE REVIEW

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 1]
```diff
@@ -65,23 +65,49 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
-    try {
-        error = JSON.parse(response).error;
+/** Empirically-based model of error response inner error from ODSP */
+interface OdspErrorResponseInnerError {
+    code: string;
+    innerError?: OdspErrorResponseInnerError
+}
+
+/** Empirically-based model of error responses from ODSP */
+interface OdspErrorResponse {
```
Comment (markfields) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:77]: @GaryWilber - Is this format documented?

**Comment by markfields** [comment]
Here's a canonical example from the logs that illustrates the types I'm adding:

```
{
  "error": {
    "code": "unauthenticated",
    "innerError": {
      "code": "invalidSignature"
    },
    "message": "Token contains invalid signature."
  }
}
```

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 2]
```diff
@@ -65,23 +65,49 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
+    catch(e) {}
+    return { success: false };
+}
 
+function parseFacetCodes(errorResponse: OdspErrorResponse): string[] {
+    const stack: string[] = [];
+    let error: OdspErrorResponseInnerError | undefined = errorResponse.error;
     // eslint-disable-next-line no-null/no-null
     while (typeof error === "object" && error !== null) {
-        if (error.code !== undefined) {
```
Comment (markfields) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:82]: I checked and in 100% of the cases in the logs (30 day retention or so), if `message` is set on the response error then `code` is as well.  Hence the type guard above checking for both, and we can confidently remove this additional check.

**Comment by markfields** [comment]
@chensixx, @jatgarg  -- Could either of you take a look before I leave tomorrow please?  Thanks!

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 2 above)
Comment (markfields) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:82]: Reverted it back after seeing UTs handle it, it's no big deal I think.

**Review by jatgarg — APPROVED** [review]
:shipit:

**Comment by vladsud** [comment]
@markfields , can we please note introduce new names like odspResponseErrorMessage?
I'm of an opinion 99.9% of the queries we do should include eventName, error, reason, sequenceNumber, and that should be enough in majority of the cases to tell developer what's going on. When I need to start looking at each individual event and what payload it has, it slows down analyzes like 100x.
There is a place for unique names, but it should be more of an exception when we can use more standartized names.

**Comment by markfields** [comment]
@vladsud Yeah it pains me to do this.  Goal will be to put that thing in `message` once the existing `message` payload gets moved to `fluidErrorCode`, but that is a much more difficult refactor.  I prioritized doing _something_ here to get rid of logging the raw response payload.

That said, good suggestion for using `response`, I'll make that change.  In fact... maybe I will leave the full response intact if it looks (vaguely) like we expect.  Then when more ready to adapt these logs more completely we can reconsider where to put what info.

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 3]
```diff
@@ -65,18 +65,46 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
- }
+}
 
-export function parseFacetCodes(response: string): string[] {
-    const stack: string[] = [];
-    let error;
-    try {
-        error = JSON.parse(response).error;
+/** Empirically-based model of error response inner error from ODSP */
+export interface OdspErrorResponseInnerError {
```
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:71]: Consider moving it to contracts.ts

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 4]
```diff
@@ -65,18 +65,46 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
+/** Empirically-based type guard for error responses from ODSP */
+function isOdspErrorResponse(x: any): x is OdspErrorResponse {
+    const error = x?.error;
+    return typeof(error?.message) === "string" &&
+        (error?.code === undefined || typeof(error?.code) === "string");
+}
+
+export function tryParseErrorResponse(
+    response: string | undefined,
+): { success: true, errorResponse: OdspErrorResponse } | { success: false } {
```
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:92]: It is much easier to work with  OdspErrorResponse  | undefined than with types like that.

**Review by vladsud — APPROVED** [review]

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 4 above)
Comment (markfields) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:92]: Yeah that's a fair critique. I'd consider cleaning up in a subsequent PR after my leave of absence.

**Pull request merged**